heron/healthmgr/src/java/com/twitter/heron/healthmgr/detectors/SkewDetector.java

```diff
@@ -15,53 +15,100 @@
-    this.sensor = sensor;
+  SkewDetector(double skewRatio, BaseSensor.MetricName metricName, BaseDetector.SymptomType
+      symptomType) {
     this.skewRatio = skewRatio;
-    this.symptomName = symptom;
+    this.metricName = metricName.text();
+    this.symptomType = symptomType;
   }
 
   /**
-   * Detects components experiencing data skew, instances with vastly different execute counts.
+   * Detects components experiencing skew on a specific metric
    *
-   * @return A collection of affected components
+   * @return At most two symptoms corresponding to each affected component -- one for positive skew
+   * and one for negative skew
    */
   @Override
-  public List<Symptom> detect() {
-    ArrayList<Symptom> result = new ArrayList<>();
-
-    Map<String, ComponentMetrics> metrics = sensor.get();
-    for (ComponentMetrics compMetrics : metrics.values()) {
-      ComponentMetricsHelper compStats = new ComponentMetricsHelper(compMetrics);
-      MetricsStats stats = compStats.computeMinMaxStats(sensor.getMetricName());
-      if (stats.getMetricMax() > skewRatio * stats.getMetricMin()) {
-        LOG.info(String.format("Detected skew for %s, min = %f, max = %f",
-            compMetrics.getName(), stats.getMetricMin(), stats.getMetricMax()));
-        result.add(new Symptom(symptomName.text(), compMetrics));
+  public Collection<Symptom> detect(Collection<Measurement> measurements) {
+    Collection<Symptom> result = new ArrayList<>();
+
+    MeasurementsTable metrics = MeasurementsTable.of(measurements).type(metricName);
+    Instant now = context.checkpoint();
+    for (String component : metrics.uniqueComponents()) {
+      Set<String> addresses = new HashSet<>();
+      Set<String> positiveAddresses = new HashSet<>();
+      Set<String> negativeAddresses = new HashSet<>();
+
+      double componentMax = getMaxOfAverage(metrics.component(component));
+      double componentMin = getMinOfAverage(metrics.component(component));
+      if (componentMax > skewRatio * componentMin) {
+        //there is skew
+        addresses.add(component);
+        result.add(new Symptom(symptomType.text(), now, addresses));
+
+        for (String instance : metrics.component(component).uniqueInstances()) {
+          if (metrics.instance(instance).mean() >= 0.90 * componentMax) {
+            positiveAddresses.add(instance);
+          }
+          if (metrics.instance(instance).mean() <= 1.10 * componentMin) {
+            negativeAddresses.add(instance);
+          }
+        }
+
+        if (!positiveAddresses.isEmpty()) {
+          result.add(new Symptom("POSITIVE " + symptomType.text(), now, positiveAddresses));
+        }
+        if (!negativeAddresses.isEmpty()) {
+          result.add(new Symptom("NEGATIVE " + symptomType.text(), now, negativeAddresses));
+        }
       }
-    }
 
+    }
     return result;
   }
+
+  @VisibleForTesting
+  double getMaxOfAverage(MeasurementsTable table) {
```

Comment:
why not put the two helper methods in `MeasurementsTable`

Reply:
Agree. I have created an issue in Dhalion repo. I am planning to incorporate the related refactoring in an independent PR. Let me know if that is ok.

Reply:
sgtm